src/textual/walk.py

```diff
@@ -50,29 +50,37 @@ def walk_depth_first(
 
     Args:
         root: The root note (starting point).
-        filter_type: Optional DOMNode subclass to filter by, or ``None`` for no filter.
+        filter_type: Optional DOMNode subclass to filter by, or `None` for no filter.
         with_root: Include the root in the walk.
 
     Returns:
-        An iterable of DOMNodes, or the type specified in ``filter_type``.
+        An iterable of DOMNodes, or the type specified in `filter_type`.
     """
-    from textual.dom import DOMNode
-
     stack: list[Iterator[DOMNode]] = [iter(root.children)]
     pop = stack.pop
     push = stack.append
-    check_type = filter_type or DOMNode
 
-    if with_root and isinstance(root, check_type):
-        yield root
-    while stack:
-        if (node := next(stack[-1], None)) is None:
-            pop()
-        else:
-            if isinstance(node, check_type):
+    if filter_type is None:
+        if with_root:
+            yield root
+        while stack:
+            if (node := next(stack[-1], None)) is None:
+                pop()
+            else:
                 yield node
-            if children := node._nodes:
-                push(iter(children))
+                if children := node._nodes:
+                    push(iter(children))
+    else:
+        if with_root and isinstance(root, filter_type):
+            yield root
+        while stack:
+            if (node := next(stack[-1], None)) is None:
+                pop()
+            else:
+                if isinstance(node, filter_type):
```

Comment:
I feel like writing 'if filter_type and isinstance(...)' here would have the same overall result and would only be a single check of a Boolean value, but would allow us to keep the single loop.

Reply:
It would still be a check per-node. Which we can avoid.

I did say it was OTT!